rs/ethereum/ledger-suite-orchestrator/src/state/mod.rs

```diff
@@ -213,7 +215,100 @@ pub type GitCommitHash = Hash<GIT_COMMIT_HASH_LENGTH>;
 
 #[derive(Clone, PartialEq, Debug, Default, Deserialize, Serialize)]
 pub struct ManagedCanisters {
+    /// Canisters for an ERC-20 token
     canisters: BTreeMap<Erc20Token, Canisters>,
+    /// Canisters for non ERC-20 token, identified by their token symbol
+    #[serde(default)]
+    other_canisters: BTreeMap<TokenSymbol, Canisters>,
+}
+
+impl ManagedCanisters {
+    pub fn find_by_id(&self, token_id: &TokenId) -> Option<&Canisters> {
+        match token_id {
+            TokenId::Erc20(contract) => self.canisters.get(contract),
+            TokenId::Other(symbol) => self.other_canisters.get(symbol),
+        }
+    }
+
+    pub fn get_mut(&mut self, token_id: &TokenId) -> Option<&mut Canisters> {
+        match token_id {
+            TokenId::Erc20(contract) => self.canisters.get_mut(contract),
+            TokenId::Other(symbol) => self.other_canisters.get_mut(symbol),
+        }
+    }
+
+    pub fn insert_once(&mut self, token_id: TokenId, canisters: Canisters) {
+        assert_eq!(
+            self.find_by_id(&token_id),
+            None,
+            "BUG: token {:?} is already managed",
+            token_id
+        );
+        let previous_element = match token_id {
+            TokenId::Erc20(contract) => self.canisters.insert(contract, canisters),
+            TokenId::Other(token_symbol) => self.other_canisters.insert(token_symbol, canisters),
+        };
+        assert_eq!(previous_element, None);
+    }
+
+    pub fn all_canisters_iter(&self) -> impl Iterator<Item = (TokenId, &Canisters)> {
+        self.canisters
+            .iter()
+            .map(|(key, value)| (TokenId::from(key.clone()), value))
+            .chain(
+                self.other_canisters
+                    .iter()
+                    .map(|(key, value)| (TokenId::from(key.clone()), value)),
+            )
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Deserialize, Serialize)]
+pub struct TokenSymbol(String);
+
+impl Display for TokenSymbol {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<String> for TokenSymbol {
+    fn from(value: String) -> Self {
+        Self(value)
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Deserialize, Serialize)]
+pub enum TokenId {
+    Erc20(Erc20Token),
+    Other(TokenSymbol),
+}
+impl TokenId {
+    pub fn into_erc20_unchecked(self) -> Erc20Token {
+        match self {
+            TokenId::Erc20(token) => token,
+            TokenId::Other(_) => panic!("BUG: token id is not an ERC-20 token"),
```

Comment:
Would it make sense to log the token ID here (and in the function below)?

Reply:
good idea, done.